bpftool: Allow to select sections and filter probes

vadorovsky opened the pull request:
This pull request introduces two changes in `bpftool feature` syntax:

- ability to select a specific section of probes to execute and print
- ability to filter in and filter out probes with regular expressions to execute and print

The only exception in terms of executing is the `syscall_config` probe, which is always executed, but if the other section was provided as an argument, `syscall_config` check will perform silently without printing and exit bpftool if the bpf() syscall is not available (because in that case running any probe has no sense).

Syntax of the `bpftool feature` after this change is:

```
bpftool feature probe [COMPONENT] section [SECTION] [filter_in PATTERN] [filter_out PATTERN] [macros [prefix PREFIX]]
COMPONENT := { kernel | dev NAME }
SECTION := { all | system_config | syscall_config | program_types | map_types | helpers | misc }
```

Please review per commit.

Fixes cilium/cilium#10048

vadorovsky committed:
bpftool: Move out sections to separate functions

Remove all calls of print_end_then_start_section function and for loops
out from the do_probe function. Instead, provide separate functions for
each section (like i.e. section_helpers) which are called in do_probe.

Next changes are going to allow probing chosen subsets of features.
This change will make them more readable.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpftool: Allow to select a specific section to probe

This change introduces a new positional argument "section" which takes
the following arguments:

- system_config
- syscall_config
- program_types
- map_types
- helpers
- misc

If "section" argument is defined, only that particular section is going
to be probed and printed. The only section which is always going to be
probed is "syscall_config", but if the other section was provided as an
argument, "syscall_config" check will perform silently without printing
and exit bpftool if the bpf() syscall is not available (because in that
case running any probe has no sense).

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpftool: Add arguments for filtering in and filtering out probes

Add positional arguments "filter_in" and "filter_out" to the feature
command of bpftool. If "filter_in" is defined, bpftool is going to
perform and print only checks which match the given pattern. If
"filter_out" is defined, bpftool will not perform and print checks which
match the given pattern.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky commented on tools/bpf/bpftool/feature.c:
```diff
@@ -57,6 +58,57 @@ static void uppercase(char *str, size_t len)
+	/* Do not probe if filter_in was defined and string does not match
+	 * against the pattern.
+	 */
+	if (filter_in != NULL) {
```
I probably need to write some macro for that.

qmonnet replied:
What do you mean?

vadorovsky replied:
Lines 72-86 repeat two almost identical blocks of code. But now I have a better solution in mind how to change that.

vadorovsky commented:
The only problem with this PR and the reason why it's WIP - glibc regex library works weird for me. Filtering out those two patterns separately (`bpftool feature kernel filter_out trace` and `bpftool feature kernel filter_out perf`) works perfect. Trying to filter them out all together (I tried `bpftool feature kernel filter_out "trace|perf"`, `bpftool feature kernel filter_out "(trace|perf)"`, `bpftool feature kernel filter_out ".*(trace|perf).*`) doesn't work at all and none of patterns get filtered out.

vadorovsky commented on tools/bpf/bpftool/feature.c:
```diff
@@ -629,6 +873,67 @@ static int do_probe(int argc, char **argv)
+				return -1;
+			filter_in_raw = GET_ARG();
+
+			filter_in = malloc(sizeof(regex_t));
```
And here I probably need to write some macro too.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -658,94 +769,18 @@ static int do_probe(int argc, char **argv)
+	section_misc(define_prefix, ifindex);
 
 exit_close_json:
 	if (json_output) {
```
Nit: Curly braces not needed, please remove.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -658,94 +769,18 @@ static int do_probe(int argc, char **argv)
-
-	if (!probe_bpf_syscall(define_prefix))
+	section_system_config(target, define_prefix);
+	if (!section_syscall_config(define_prefix)) {
```
Nit: Curly braces are no longer necessary here, please remove them.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -608,35 +612,42 @@ section_system_config(enum probe_component target, const char *define_prefix)
 			    "/*** System call availability ***/",
 			    define_prefix);
-	res = probe_bpf_syscall(define_prefix);
+	res = probe_bpf_syscall(print_syscall_config, define_prefix);
```
I'd rather have just one call to `probe_bpf_syscall()`, and make the `print_*_section()` conditional, like this:
```
if (print_syscall_config)
    print_start_section(...);
res = probe_bpf_syscall(...);
if (print_syscall_config)
    print_end_section();
return res;
```
... but this is just a personal preference I suppose, so fine either way.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -699,8 +710,25 @@ static void section_misc(const char *define_prefix, __u32 ifindex)
 	const char *define_prefix = NULL;
+	bool check_system_config = false;
+	/* Program types probes are needed if helper probes are going to be
+	 * performed. Therefore we should diffirentiate between checking and
```
s/diffirentiate/differentiate/

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -584,13 +578,130 @@ probe_large_insn_limit(const char *define_prefix, __u32 ifindex)
+
+static void
+section_program_types(bool *supported_types, const char *define_prefix,
+		     __u32 ifindex)
```
Nit: checkpatch with `--strict` complains about the alignment of `__u32 ifindex`. Not a big issue since you change it in later patches.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -699,8 +710,25 @@ static void section_misc(const char *define_prefix, __u32 ifindex)
 static int do_probe(int argc, char **argv)
 {
 	enum probe_component target = COMPONENT_UNSPEC;
+	/* Syscall probe is always performed, because performing any other
+	 * checks without bpf() syscall does not make sense and the program
+	 * should exit.
+	 */
```
Nit: We do perform other checks (first section, e.g. kernel config options). Maybe something along “... because performing other checks relying on the bpf() system call does not make sense...”?

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -740,6 +768,35 @@ static int do_probe(int argc, char **argv)
+			check_section = true;
+			NEXT_ARG();
+			if (is_prefix(*argv, "all")) {
+				check_section = false;
```
Your implementation mostly supports using the `section` keyword multiple time, but please compare the output for:
- `bpftool feature probe section program_types section all`
- `bpftool feature probe section all section program_types`

Because you reset `check_section` to `false`, the previous selection is reset. Maybe add an additonal `check_all`, and set all the rest to true at the end if `section all` is encountered?

vadorovsky replied:
I think that maybe we should not allow to mix `all` with requesting any concrete sections? That doesn't really make sense - to request `all` and then request i.e. `program_types`.

qmonnet replied:
Or maybe just remove `section all` entirely (just don't use `section` in that case)?

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -740,6 +768,35 @@ static int do_probe(int argc, char **argv)
+				/* When helpers probes are requested, program
+				 * types probes have to be performed, but they
+				 * will be not printed.
+				 */
```
s/will not be printed/may not be printed/

(It depends on what sections were required by the user, of course)

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -740,6 +768,35 @@ static int do_probe(int argc, char **argv)
+			} else if (is_prefix(*argv, "misc")) {
+				check_misc = true;
+			} else {
+				p_err("unrecognized section '%s'", *argv);
```
Could be helpful to list available section names here in the error message.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -764,20 +821,37 @@ static int do_probe(int argc, char **argv)
 	}
 
+	/* Perform all checks if specific section check was not requested. */
+	if (!check_section) {
```
(With regard to earlier comment: This could be `if (!check_section || check_all)`)

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -764,20 +821,37 @@ static int do_probe(int argc, char **argv)
-	if (!section_syscall_config(define_prefix)) {
+	if (check_system_config)
+		section_system_config(target, define_prefix);
+	if (!section_syscall_config(print_syscall_config, define_prefix)) {
```
Why would we run syscall checks if user only wants system config probes? :)

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -57,6 +58,57 @@ static void uppercase(char *str, size_t len)
+			break;
+		case REG_NOMATCH:
+			return false;
+			break;
```
`break` unnecessary after `return`

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -57,6 +58,57 @@ static void uppercase(char *str, size_t len)
+	/* Do not probe if filter_in was defined and string does not match
+	 * against the pattern.
+	 */
+	if (filter_in != NULL) {
```
`if (filter_in)`

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -57,6 +58,57 @@ static void uppercase(char *str, size_t len)
+	/* Do not probe if filter_out was defined and string matches against the
+	 * pattern.
+	 */
+	if (filter_out != NULL) {
```
`if (filter_out)`

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -57,6 +58,57 @@ static void uppercase(char *str, size_t len)
+	if (filter_in != NULL) {
+		ret = regexec(filter_in, name, 0, NULL, 0);
+		switch (ret) {
+		case 0:
+			break;
```
So as I understand, we can have:
- `filter_in trace filter_out trace_printk`

But we cannot have:
- `filter_out trace filter_in trace_prink`

I mean, we _can_ use it, but it will not exclude all items with `*trace*` but keep only `bpf_trace_prink()`, because `filter_in` elements apply to the initial selection while `filter_out` is also applied to elements kept with `filter_in`.

I'm ok with that, but this is something that must be documented when you update the man page (possibly with better words than mine, actually :) ).

vadorovsky replied:
I mentioned that in the man page.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -797,6 +902,38 @@ static int do_probe(int argc, char **argv)
+			NEXT_ARG();
+			if (!REQ_ARGS(1))
+				return -1;
+			filter_in_raw = GET_ARG();
```
So `filter_in` and `filter_out` can be used only once on the command line. Please do not forget to mention it in the man page (or even bail out if it is passed multiple time, i.e. if `filter_in_raw` is already set).

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -724,13 +822,20 @@ static int do_probe(int argc, char **argv)
 	bool check_program_types = false;
 	bool print_program_types = false;
+	const char *filter_out_raw= NULL;
+	const char *filter_in_raw= NULL;
```
Missing space before `=`. Don't forget to run checkpatch (I'd recommend the `--strict` option) :)

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -838,28 +975,37 @@ static int do_probe(int argc, char **argv)
 	}
 
-	return 0;
+cleanup:
```
Arguments need not be passed in a specific order. So if you have `filter_in <something>` and then `prefix foo prefix bar`, the function will `return -1`, without `free`ing `filter_in` and `filter_out`. If you want to exit cleanly, you'll have to replace most `return -1;` in the function with `ret = -1; goto cleanup`.

vadorovsky replied:
Good point. I replaced `return -1;` with `ret = -1; goto cleanup` in every argument, just to be sure.

qmonnet reviewed (commented):
Looks good overall. There are some minor improvements to bring here and there, see the comments.

I had not thought of the prog types probes being required for probing the helpers. It is unfortunate because it complicates things for the helpers, and makes the code less readable by passing `print_program_types` everywhere. Your approach is good I think, although we can have some weird stuff like `bpftool feature probe filter_out prog` which says that no helper is supported whatsoever. Maybe we should still probe for the prog types internally even if they are filtered out? 

Two additional generic comments, not sure if worth addressing, mostly recommendations for the future:
- I would keep the most important arguments at the beginning of the function definitions when adding new ones (e.g. `probe_prog_type(type, print_program_types)` instead of `probe_prog_type(print_program_types, type)`.
- I would try to use verbs in function names (thinking of `section_program_types` and similar).

Regarding the regex library, I managed to get multiple matches with e.g.:
```
bpftool feature probe filter_in "\(trace\|perf\)"
```

Also don't forget bash completion along with the man page :)

vadorovsky commented:
>Maybe we should still probe for the prog types internally even if they are filtered out?

I will try that. If it will not generate any unwanted dmesg entries, then that's a good idea.

vadorovsky commented:
@qmonnet Your comments should be addressed now. Please take a look again.

The bpftool command which is showing all probes we need, but does not pollute dmesg, is:

```
bpftool feature probe filter_out "\(trace\|perf\|write_user\)"
```

I will make a PR in Cilium with that change.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -740,6 +767,35 @@ static int do_probe(int argc, char **argv)
+				check_section = false;
+			} else if (is_prefix(*argv, "system_config")) {
+				check_system_config = true;
+			} else if (is_prefix(*argv, "syscall_config")) {
```
Note: potential conflict on prefixing (`bpftool f p section sys` will default to `system_config`). Same thing for `map_types` and `misc`. I don't think this is an issue, though.

vadorovsky replied:
True, maybe we should just use `strcmp`?

qmonnet replied:
It was a remark, but I think I would just keep your change. Prefixing is useful. We could error if the prefix is ambiguous (I think I implemented that once in the past in bpftool), but this is probably not worth the hastle. Worst case: user gets probes for the wrong section, they realise it and fix their command. Not as if we had a risk of doing unexpected modifications to the system, so let's just stick to your version.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -740,6 +767,35 @@ static int do_probe(int argc, char **argv)
+				 * will be not printed.
+				 */
+				check_program_types = true;
+				print_program_types = false;
```
`print_program_types` defaults to `false`, you should not change it here. Especially because if user types `bpftool f p section program_types section helpers`, selecting the helpers will disable printing for program type probes, although they were required by the user.

vadorovsky replied:
done

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -764,19 +820,36 @@ static int do_probe(int argc, char **argv)
 
+	/* Perform all checks if specific section check was not requested. */
+	if (!check_section) {
+		check_system_config = true;
```
Nit: reverse Christmas tree order
(I think you were good before adding the missing space before the `=` on the line below :). Please update)

vadorovsky replied:
Done.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -754,23 +857,37 @@ static int do_probe(int argc, char **argv)
+				    check_map_types ||
+				    check_helpers ||
+				    check_misc) {
+					p_err("cannot mix 'all' with requesting other sections");
```
This should be moved to the previous commit.

vadorovsky replied:
I removed `all` as you proposed in the other comment.

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -784,39 +901,92 @@ static int do_probe(int argc, char **argv)
 				/* When helpers probes are requested, program
 				 * types probes have to be performed, but they
-				 * will be not printed.
+				 * may not be printed.
```
This also belongs to the previous commit.

vadorovsky replied:
done

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -784,39 +901,92 @@ static int do_probe(int argc, char **argv)
-				p_err("unrecognized section '%s'", *argv);
-				return -1;
+				p_err("unrecognized section '%s', available sections: system_config, "
+				      "syscall_config, program_tyoes, map_types, helpers, misc", *argv);
```
Typo on `program_tyoes`

vadorovsky replied:
done

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -727,9 +824,14 @@ static int do_probe(int argc, char **argv)
 	__u32 ifindex = 0;
 	char *ifname;
+	int reg_ret;
+	int ret = 0;
```
Note: Instead of `ret = 0`, we could have used `err = -1`, and `return err;` at the end of the function (with `err = 0` in case of success). We would have avoided all the `ret = -1` before the `goto cleanup`.

But it's fine this way too, no need to change :).

qmonnet commented on tools/bpf/bpftool/feature.c:
```diff
@@ -431,37 +487,54 @@ static void probe_kernel_image_config(void)
+	sprintf(plain_desc, "%s%s", plain_comment, prog_type_name[prog_type]);
+
+	if (!check_filters(feat_name, filter_in, filter_out))
+		return;
```
No. This will remove the possibility to probe helpers if program types are filtered out.

This check should be moved lower down, before the call to `print_bool_feature()`.

vadorovsky replied:
done

qmonnet reviewed (commented):
Thanks for the changes! There are a few remaining nits and minor issues, to be fixed before upstream submission.

I do not believe they are blocking for merging this PR to our local branch.

vadorovsky committed:
bpftool: Update documentation of "bpftool feature" command

Update documentation of "bpftool feature" command with information about
new arguments: "section", "filter_in" and "filter_out".

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky commented:
I pushed also a man page change. I still need to add bash completion. But that shouldn't be a blocker for Cilium fork.

qmonnet commented on tools/bpf/bpftool/Documentation/bpftool-feature.rst:
```diff
@@ -19,19 +19,42 @@ SYNOPSIS
 		  functions availability, and more.
 
+		  If the **section** keyword is passed, only the specified
+		  probes section will be checked and printed. The only section
```
s/section whis is probed/probe which is performed/ ?

qmonnet commented on tools/bpf/bpftool/Documentation/bpftool-feature.rst:
```diff
@@ -19,19 +19,42 @@ SYNOPSIS
+		  which is always going to be probed is **syscall_config**,
+		  but if the other section was provided as an argument,
+		  **syscall_config** check will perform silently without
+		  printing the result and bpftool will exit if the bpf()
```
s/`bpf()`/`**bpf()**`/

(here and below)

qmonnet commented on tools/bpf/bpftool/Documentation/bpftool-feature.rst:
```diff
@@ -19,19 +19,42 @@ SYNOPSIS
+		  but if the other section was provided as an argument,
+		  **syscall_config** check will perform silently without
+		  printing the result and bpftool will exit if the bpf()
+		  syscall is not abailable (because in that case performing
```
s/abailable/available/

qmonnet commented on tools/bpf/bpftool/Documentation/bpftool-feature.rst:
```diff
@@ -19,19 +19,42 @@ SYNOPSIS
+
+		  **filter_in** is executed before **filter_out** which means
+		  that **filter_out** is always applied only on probes
+		  selected by **filter_in** if both arguments are used together.
```
“`Please refer to the **regex**\ (7) man page for details on the syntax for *PATTERN*.`”

qmonnet commented on tools/bpf/bpftool/Documentation/bpftool-feature.rst:
```diff
@@ -19,19 +19,42 @@ SYNOPSIS
 ================
 
-|	**bpftool** **feature probe** [*COMPONENT*] [**macros** [**prefix** *PREFIX*]]
+|	**bpftool** **feature probe** [*COMPONENT*] [**section** [*SECTION*]] [**filter_in** *PATTERN*] [**filter_out** *PATTERN*] [**macros** [**prefix** *PREFIX*]]
```
If `section` is passed, then `SECTION` is not optional. We should have `[**section** *SECTION*]` instead of `[**section** [*SECTION*]]` I think? (Here and below)

qmonnet reviewed (approved):
Thanks for the changes, looks good now!

A few nits on the man page :) But good overall.
Another nit on the code: don't forget to update the help message in `do_help()`, at the end of feature.c, before submitting upstream.

vadorovsky commented:
Thanks for review! I will merge this one and submit patches upstream after addressing last comments carefully.
